### PR TITLE
remove const_cast and refactor ngraph engine code 

### PR DESCRIPTION
**Pull request by baojun-nervana**
1. close a TODO to remove the const_cast. This is to avoid altering the progdesc. Instead of substituting ngraph_engine op in progdesc, it is changed back to scan the op list after creation and replace the op sublist with ngraph_engine (the original way).

2. Refactor the engine code to reduce the overhead. The idea is to minimize the repetitives iteration to iteration. If the graph has no change, the NgraphEngine::Run will grab the data and run the computation.
    - an "engine_key" was added to verify if there is a graph change.
    - the repetitive info for NgraphEngine::Run will be cached

### CODE REVIEW

**Comment by baojun-nervana** [comment]
[23:45:10]	+ '[' ']'
[23:45:10]	++ git diff -U0 upstream/develop
[23:45:10]	++ grep -o -m 1 const_cast
[23:45:10]	+ HAS_CONST_CAST=const_cast
[23:45:10]	+ '[' const_cast ']'
[23:45:10]	+ '[' 15925 '!=' '' ']'
[23:45:10]	++ curl -H 'Authorization: token 95a329e820eeae12c4c5c8a0b6e3e39654dc01a4' 'https://api.github.com/repos/PaddlePaddle/Paddle/pulls/15925/reviews?per_page=10000'
[23:45:10]	++ python /paddle/tools/check_pr_approval.py 1 2887803
[23:45:10]	  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[23:45:10]	                                 Dload  Upload   Total   Spent    Left  Speed
[23:45:10]	
[23:45:11]	  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[23:45:12]	  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[23:45:12]	100     5  100     5    0     0      3      0  0:00:01  0:00:01 --:--:--     3
[23:45:12]	+ APPROVALS=FALSE
[23:45:12]	+ echo 'current pr 15925 got approvals: FALSE'
[23:45:12]	+ '[' FALSE == FALSE ']'
[23:45:12]	+ echo 'You must have panyx0718 approval for the const_cast'
[23:45:12]	+ exit 1
[23:45:19]	Process exited with code 1

File: paddle/fluid/framework/executor.cc
Comment (tensor-tang) [inline paddle/fluid/framework/executor.cc]: is that necessary include these?

**Review by tensor-tang — COMMENTED** [review]
I haven't digger deep into this code, but since we are reviewing the design doc inside, maybe you could add this conception as well, especially about the cache method you used.

At the most possibility, you should use PASS to modify the program.

File: paddle/fluid/framework/executor.cc
Comment (baojun-nervana) [reply paddle/fluid/framework/executor.cc]: @tensor-tang It is the cpplint requirement from pre-commit to add those. It seems cpp-lint is on for CI. I have to add those to pass CI. 

I can add the cache to the design doc. 

We will support pass for ngraph enabling, but I think we will still need the current ngraph triggering in executor to support unittest.

Comment (tensor-tang) [reply paddle/fluid/framework/executor.cc]: > I have to add those to pass CI.

OK, got it. we have updated the cpp-lint in CI.

**Comment by tensor-tang** [comment]
Please fix the errors in CI. 
BTW, the cpplint version is 1.4.4 in CI/PR, you can check style locally.

File: paddle/fluid/operators/ngraph/ngraph_engine.h
Comment (tensor-tang) [inline paddle/fluid/operators/ngraph/ngraph_engine.h:15]: use `pragma once` instead.

**Review by tensor-tang — APPROVED** [review]
LGTM

File: paddle/fluid/operators/ngraph/ngraph_engine.h
Comment (baojun-nervana) [reply paddle/fluid/operators/ngraph/ngraph_engine.h:15]: will do. Maybe later PR (not sure if it will dismiss your approval if I do it in this PR)

**Review by panyx0718 — APPROVED** [review]
LG const_cast removal

File: paddle/fluid/operators/ngraph/ngraph_engine.h
Comment (tensor-tang) [reply paddle/fluid/operators/ngraph/ngraph_engine.h:15]: You can address this in #15752 .

**Pull request merged**
